### PR TITLE
Added TestDataGenerator and tests

### PR DESCRIPTION
**Pull request by DanielMcCrystal**
On the admin page, an admin can specify a number of fake users, conversations, and messages to add to the site. These entities appear in the conversation page and in the site statistics. The admin can manually clear all the generated test data, but all test data gets removed anyway when the server restarts, as it is never actually written to the datastore.

### CODE REVIEW

File: src/main/java/codeu/model/store/persistence/PersistentStorageAgent.java [hunk 1]
```diff
@@ -103,4 +108,20 @@ public void writeThrough(Conversation conversation) {
   public void writeThrough(Message message) {
     persistentDataStore.writeThrough(message);
   }
+
+  public void deleteEntity(UUID id) {
+    //Key key = KeyFactory.generateKey(id.toString());
```
Comment (kyrabusser) [inline src/main/java/codeu/model/store/persistence/PersistentStorageAgent.java]: Why are these two methods commented out?

File: src/test/java/codeu/controller/TestDataGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    mockMessageStore = MessageStore.getTestInstance(mockPersistentStorageAgent);
+    mockTestDataGenerator.setMessageStore(mockMessageStore);
+
+  }
+
+  @Test
+  public void testAddTestUsers() {
+
+    int numUsersBefore = mockUserStore.getAllUsers().size();
```
Comment (kyrabusser) [inline src/test/java/codeu/controller/TestDataGeneratorTest.java]: The tests should all run independently from a known starting state, so you should be able to replace all of these "before" numbers with 0s (and inline them as appropriate).

If you'd like to also test that adding test entities can be done multiple times or when there's already data in the DB, that can be tested explicitly with multiple calls or setting a new return value for mockPersistentStorageAgent.

File: src/main/java/codeu/controller/TestDataGenerator.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  void setMessageStore(MessageStore messageStore) {
+    this.messageStore = messageStore;
+  }
+  public MessageStore getMessageStore() {
+    return messageStore;
+  }
+
+  void addTestData() {
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/TestDataGenerator.java]: Is this version of the method ever used?

**Review by kyrabusser — COMMENTED** [review]
This is pretty sweet. :+1:

Do you have screenshots (especially of the slick JS button)?

And what happens if a real user interacts with the volatile test data (for example, adds a message to a test conversation)? Are there any edge cases to consider there?

File: src/main/java/codeu/model/store/persistence/PersistentStorageAgent.java (hunk 1 above)
Comment (DanielMcCrystal) [reply src/main/java/codeu/model/store/persistence/PersistentStorageAgent.java]: Whoops, I forgot I left those in. Originally I was figuring out how to delete the test data from the datastore, before I realized it would be better to just never write it in the first place.

**Comment by DanielMcCrystal** [comment]
Here are some screenshots of the JS elements.
screen shot 2018-06-18 at 3 28 25 pm
screen shot 2018-06-18 at 3 28 33 pm

**Comment by kyrabusser** [comment]
Hey Daniel! Can you write up a summary for the overall strategy for the temporary test data? I see you've started modifying the User model and session state, but that feels like it goes beyond the original scope of this specific PR. I think that if we can think through and write up what we want the final goal to be, it will be easier to split up into manageable steps. Sound good?

**Comment by DanielMcCrystal** [comment]
I wanted the test data to be real data (otherwise it wouldn't really be testing the functionality of the site). But Kyra brought up a good point, which is that real users interacting with test conversations would cause problems. To solve this, I needed to make sure only admins could see the test data. And I also needed to make sure that no user, admin or otherwise, could contribute to test conversations. I changed the User model so that I could determine the admin status of a user outside of the admin servlet. This was necessary for the above features, but was also an overdue fix to a temporary workaround (hardcoded ArrayList of admin usernames).

File: src/main/java/codeu/controller/ServerStartupListener.java [hunk 4]
```diff
@@ -25,6 +32,13 @@
   public void contextInitialized(ServletContextEvent sce) {
     try {
       List<User> users = PersistentStorageAgent.getInstance().loadUsers();
+
+      String adminPassword = "admin";
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/ServerStartupListener.java:36]: Nit: I would extract this hardcoded username and password into constants--make them private static final and put them at the top of the file. That way they're easier to find by looking at the beginning of this class, rather than buried in the method implementation. (https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names)

File: src/main/java/codeu/controller/ServerStartupListener.java [hunk 5]
```diff
@@ -25,6 +32,13 @@
   public void contextInitialized(ServletContextEvent sce) {
     try {
       List<User> users = PersistentStorageAgent.getInstance().loadUsers();
+
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/ServerStartupListener.java:35]: Update the comment on this contextInitialized method, since we're doing more than just loading data now.

File: src/main/java/codeu/controller/AdminServlet.java [hunk 6]
```diff
@@ -149,18 +136,60 @@ void computeAdminStats(HttpServletRequest request) {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
-    throws IOException, ServletException {
+      throws IOException, ServletException {
+
+    if (request.getSession().getAttribute("isAdmin").equals(Boolean.TRUE)) {
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/AdminServlet.java:141]: Oh, interesting--the .equals() method keeps you from having to cast?

Also, this will throw a NullPointerException if the "isAdmin" attribute isn't set, right? That's probably fine for the context of this project, but it's something to keep in mind for more robust applications, since "A servlet should be able to handle cases in which the client does not choose to join a session, such as when cookies are intentionally turned off.", according to the HttpSession javadoc: https://javaee.github.io/javaee-spec/javadocs/javax/servlet/http/HttpSession.html

No need to change anything, but this could be part of a bigger "product" discussion of what the experience should be like for non-logged-in users.

File: src/main/java/codeu/controller/ConversationServlet.java [hunk 7]
```diff
@@ -71,6 +72,20 @@ void setConversationStore(ConversationStore conversationStore) {
...
       throws IOException, ServletException {
     List<Conversation> conversations = conversationStore.getAllConversations();
+
+    // Filter out test data to non-admins
+    HttpSession session = request.getSession();
+    boolean adminStatus = session.getAttribute("isAdmin") != null ? (boolean) session.getAttribute("isAdmin") : false;
+
+    if (adminStatus == false) {
+      TestDataGenerator tdg = TestDataGenerator.getInstance();
+      for (int i = conversations.size() - 1; i >= 0; i--) {
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/ConversationServlet.java:82]: Ah, interesting! Iterating backwards is a cute trick for allowing you to modify the list without messing up the indicies, but I think it would be clearer to read if you make a new list for the filtered data and copy the valid conversations into it. (It took me a minute to read through the code and understand why you couldn't just use a for-each loop here.)

Java8 streams are a nice, succinct way to do filtering like this--let me know if you'd like to talk through using them.!

File: src/main/java/codeu/controller/ConversationServlet.java [hunk 8]
```diff
@@ -71,6 +72,20 @@ void setConversationStore(ConversationStore conversationStore) {
   public void doGet(HttpServletRequest request, HttpServletResponse response)
       throws IOException, ServletException {
     List<Conversation> conversations = conversationStore.getAllConversations();
+
+    // Filter out test data to non-admins
+    HttpSession session = request.getSession();
+    boolean adminStatus = session.getAttribute("isAdmin") != null ? (boolean) session.getAttribute("isAdmin") : false;
+
+    if (adminStatus == false) {
+      TestDataGenerator tdg = TestDataGenerator.getInstance();
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/ConversationServlet.java:81]: Hm, this will cause a full TestDataGenerator, including the PersistentStorageAgent, to get created in the ConversationServletTest. It would probably be better practice to set the TestGenerator the same way the datastores are set, so we can mock it out for testing.

Speaking of which, are there tests for this filtering logic?

**Review by kyrabusser — APPROVED** [review]
Apologies for the delayed review! I was hoping to give everyone on the team the chance to take a look, but I don't want you to be blocked.

Overall, you've done a really great job adding admin-specific functionality! I have some comments for readability and testing, and then I'm excited to see this merged and pushed to AppEngine. :tada: 

One larger theme that I think this PR brought up is the tension between engineering/implementation choices and big-picture/product decisions. Some options that seem like simple implementation details (like in-memory test data) can actually have pretty significant effects for the final product (like whether admins and normal users can see the same conversation list). Sometimes it's helpful to get more specific about what the "product vision" is before you begin implementing, so that the team can be on the same page when those trade-offs come up. On the other hand, you don't want to get stuck just designing without really implementing either. So different projects need to find different balances--I hope that finding that balance is an aspect of real world software engineering that CodeU can start to give you a taste of!
